From ebdc801b10273bd961f96bd890092f1ff2f0d24e Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 20 Nov 2024 01:49:30 +0100
Subject: [PATCH] [DogeCash] Add `CAuxPow::CheckAuxBlockHash` (#52)

This verifies the given `CAuxPow` commits to the `hashAuxBlock` given as parameter.

See the details in the comment.

It's a port of Dogecoin's `CAuxPow::check`.

(cherry picked from commit 71f2fa8d6fb78a3dcc72610838dbc7771743fded)

diff --git a/src/primitives/auxpow.cpp b/src/primitives/auxpow.cpp
index 9bf899ba85523a9df700131209accfb4b374e454..8a3cac7b47adde3d97227cd6baac4994283c8b16 100644
--- a/src/primitives/auxpow.cpp
+++ b/src/primitives/auxpow.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file license.txt or http://www.opensource.org/licenses/mit-license.php.
 
+#include <consensus/params.h>
 #include <hash.h>
 #include <primitives/auxpow.h>
 #include <stdexcept>
@@ -146,3 +147,62 @@ uint32_t CalcExpectedMerkleTreeIndex(uint32_t nNonce, uint32_t nChainId,
 
     return rand % (1 << merkleHeight);
 }
+
+util::Result<std::monostate>
+CAuxPow::CheckAuxBlockHash(const uint256 &hashAuxBlock, uint32_t nChainId,
+                           const Consensus::Params &params) const {
+    // Coinbase txs are always the first in a block, so nIndex must always be 0
+    if (nIndex != 0) {
+        return {{_("AuxPow nIndex must be 0")}};
+    }
+
+    // We can't merge mine our own chain, it must be a different one.
+    if (params.enforceStrictAuxPowChainId &&
+        VersionChainId(parentBlock.nVersion) == nChainId) {
+        return {{_("AuxPow parent has our chain ID")}};
+    }
+
+    // Ensure sane bounds for the chain merkle tree
+    if (vChainMerkleBranch.size() >= 31) {
+        return {{_("AuxPow chain merkle branch too long")}};
+    }
+
+    // Check that we are in the parent block merkle tree
+    if (parentBlock.hashMerkleRoot !=
+        ComputeMerkleRootForBranch(coinbaseTx->GetHash(), vMerkleBranch,
+                                   nIndex)) {
+        return {{_("AuxPow merkle root incorrect")}};
+    }
+
+    // Check that the chain merkle root is in the coinbase
+    uint256 hashRoot = ComputeMerkleRootForBranch(
+        hashAuxBlock, vChainMerkleBranch, nChainIndex);
+
+    if (coinbaseTx->vin.empty()) {
+        return {{_("AuxPow coinbase transaction missing input")}};
+    }
+
+    const CScript coinbaseScript = coinbaseTx->vin[0].scriptSig;
+
+    util::Result<ParsedAuxPowCoinbase> parsedCoinbase =
+        ParsedAuxPowCoinbase::Parse(coinbaseScript, hashRoot);
+
+    // Couldn't parse the coinbase, or some other violation
+    if (!parsedCoinbase) {
+        return {{util::ErrorString(parsedCoinbase)}};
+    }
+
+    const uint32_t merkleHeight = vChainMerkleBranch.size();
+    if (parsedCoinbase->nTreeSize != (1U << merkleHeight)) {
+        return {
+            {_("AuxPow merkle branch size does not match parent coinbase")}};
+    }
+
+    const uint32_t expectedIndex = CalcExpectedMerkleTreeIndex(
+        parsedCoinbase->nMergeMineNonce, nChainId, merkleHeight);
+    if (nChainIndex != expectedIndex) {
+        return {{_("AuxPow wrong chain index")}};
+    }
+
+    return {std::monostate()};
+}
diff --git a/src/primitives/auxpow.h b/src/primitives/auxpow.h
index d880939557791339e63d4f5f8a26fc6109cbaf4c..cb572202505aafe670d54e275ab3a5bbc5f365bc 100644
--- a/src/primitives/auxpow.h
+++ b/src/primitives/auxpow.h
@@ -10,6 +10,10 @@
 #include <primitives/transaction.h>
 #include <util/result.h>
 
+namespace Consensus {
+struct Params;
+} // namespace Consensus
+
 /** Bit that indicates a block has auxillary PoW. Bits below that are
  * interpreted as the "traditional" Bitcoin version. */
 static constexpr int32_t VERSION_AUXPOW_BIT_POS = 8;
@@ -148,6 +152,19 @@ public:
         READWRITE(obj.coinbaseTx, obj.hashBlock, obj.vMerkleBranch, obj.nIndex,
                   obj.vChainMerkleBranch, obj.nChainIndex, obj.parentBlock);
     }
+
+    /**
+     * Perform all the required AuxPow checks.
+     *
+     * Verifies the `hashAuxBlock` is correctly commited in `coinbaseTx` via a
+     * merkle tree in the scriptSig. The leaf index is given by
+     * `CalcExpectedMerkleTreeIndex`.
+     *
+     * Returns std::monostate (unit type) on success, or an error otherwise.
+     */
+    util::Result<std::monostate>
+    CheckAuxBlockHash(const uint256 &hashAuxBlock, uint32_t nChainId,
+                      const Consensus::Params &params) const;
 };
 
 #endif // BITCOIN_PRIMITIVES_AUXPOW_H
diff --git a/src/test/dogecoin_auxpow_check_tests.cpp b/src/test/dogecoin_auxpow_check_tests.cpp
index 132463b800e04a2315e3522131ad09147222bce2..bd60a79f077f2f6371ee4d1ab1ed19cb77116034 100644
--- a/src/test/dogecoin_auxpow_check_tests.cpp
+++ b/src/test/dogecoin_auxpow_check_tests.cpp
@@ -3,9 +3,11 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <primitives/auxpow.h>
+#include <span.h>
 #include <streams.h>
 #include <util/strencodings.h>
 
+#include <test/lcg.h>
 #include <test/util/setup_common.h>
 
 #include <boost/test/unit_test.hpp>
@@ -99,6 +101,10 @@ BOOST_AUTO_TEST_CASE(auxpow_block700000_test) {
                                     VersionChainId(header.nVersion),
                                     auxpow.vChainMerkleBranch.size()),
         56);
+
+    Consensus::Params params = CChainParams::Main({})->GetConsensus();
+    BOOST_CHECK(auxpow.CheckAuxBlockHash(
+        header.GetHash(), VersionChainId(header.nVersion), params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
@@ -120,6 +126,10 @@ BOOST_AUTO_TEST_CASE(auxpow_block800000_test) {
                                     VersionChainId(header.nVersion),
                                     auxpow.vChainMerkleBranch.size()),
         40);
+
+    Consensus::Params params = CChainParams::Main({})->GetConsensus();
+    BOOST_CHECK(auxpow.CheckAuxBlockHash(
+        header.GetHash(), VersionChainId(header.nVersion), params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
@@ -141,6 +151,10 @@ BOOST_AUTO_TEST_CASE(auxpow_block3000000_test) {
                                     VersionChainId(header.nVersion),
                                     auxpow.vChainMerkleBranch.size()),
         8);
+
+    Consensus::Params params = CChainParams::Main({})->GetConsensus();
+    BOOST_CHECK(auxpow.CheckAuxBlockHash(
+        header.GetHash(), VersionChainId(header.nVersion), params));
 }
 
 BOOST_AUTO_TEST_CASE(auxpow_parse_coinbase_test) {
@@ -338,4 +352,274 @@ BOOST_AUTO_TEST_CASE(CalcExpectedMerkleTreeIndex_test) {
                       1080);
 }
 
+BOOST_AUTO_TEST_CASE(CheckAuxBlockHash_test) {
+    Consensus::Params mainParams = CChainParams::Main({})->GetConsensus();
+    Consensus::Params testParams = CChainParams::TestNet({})->GetConsensus();
+    Consensus::Params regParams = CChainParams::RegTest({})->GetConsensus();
+
+    const std::vector<Consensus::Params> allParams{mainParams, testParams,
+                                                   regParams};
+
+    // nIndex must be 0
+    for (uint32_t nIndex : {1u, 2u, 100u, 0xffffffffu, 0x7fffffffu}) {
+        CAuxPow auxpow{};
+        auxpow.nIndex = nIndex;
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(uint256(), 0, params))
+                    .original,
+                "AuxPow nIndex must be 0");
+        }
+    }
+
+    // vChainMerkleBranch can at most be 30
+    for (size_t branchLen = 31; branchLen <= 100; ++branchLen) {
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.vChainMerkleBranch.resize(branchLen);
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(uint256(), 1, params))
+                    .original,
+                "AuxPow chain merkle branch too long");
+        }
+    }
+
+    // If a strict chain ID is enforced, we don't allow the parent to have the
+    // same chain ID as our chain.
+    for (uint32_t nChainId : {0u, 1u, 2u, 0xffffu, uint32_t(AUXPOW_CHAIN_ID)}) {
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.parentBlock.nVersion = MakeVersionWithChainId(nChainId, 0);
+        auxpow.vChainMerkleBranch.resize(31);
+        // Enforced on mainnet
+        BOOST_CHECK_EQUAL(ErrorString(auxpow.CheckAuxBlockHash(
+                                          uint256(), nChainId, mainParams))
+                              .original,
+                          "AuxPow parent has our chain ID");
+        // Enforced on regtest
+        BOOST_CHECK_EQUAL(ErrorString(auxpow.CheckAuxBlockHash(
+                                          uint256(), nChainId, regParams))
+                              .original,
+                          "AuxPow parent has our chain ID");
+        // Not enforced on testnet (so we fail on the merkle branch check)
+        BOOST_CHECK_EQUAL(ErrorString(auxpow.CheckAuxBlockHash(
+                                          uint256(), nChainId, testParams))
+                              .original,
+                          "AuxPow chain merkle branch too long");
+    }
+
+    // coinbaseTx is not in the parentBlock (or merkle proof incorrect)
+    for (size_t branchLen = 0; branchLen < 31; ++branchLen) {
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.coinbaseTx = MakeTransactionRef();
+        auxpow.vMerkleBranch.resize(branchLen);
+        auxpow.parentBlock.hashMerkleRoot = uint256S(
+            "123456789012345678901234567890123456789012345678901234567890abcd");
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(uint256(), 1, params))
+                    .original,
+                "AuxPow merkle root incorrect");
+        }
+    }
+
+    // Coinbase can't have no inputs
+    {
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.coinbaseTx = MakeTransactionRef();
+        auxpow.vMerkleBranch.resize(7);
+        auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+            auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(uint256(), 1, params))
+                    .original,
+                "AuxPow coinbase transaction missing input");
+        }
+    }
+
+    // Coinbase must have the required auxpow data in the scriptSig
+    {
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        CMutableTransaction tx;
+        tx.vin.resize(1);
+        auxpow.coinbaseTx = MakeTransactionRef(tx);
+        auxpow.vMerkleBranch.resize(7);
+        auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+            auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(uint256(), 1, params))
+                    .original,
+                "AuxPow missing chain merkle root in parent coinbase");
+        }
+    }
+
+    // Must set nTreeSize to 2^vChainMerkleBranch.size()
+    for (uint32_t nChainIndex = 0; nChainIndex < 128; ++nChainIndex) {
+        uint256 hashChildBlock = uint256S(
+            "123456789012345678901234567890123456789012345678901234567890abcd");
+
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.nChainIndex = nChainIndex;
+        auxpow.vChainMerkleBranch.resize(7);
+        uint256 hashChainRoot = ComputeMerkleRootForBranch(
+            hashChildBlock, auxpow.vChainMerkleBranch, nChainIndex);
+        std::reverse(hashChainRoot.begin(), hashChainRoot.end());
+
+        std::vector<uint8_t> coinbaseBytes = ToByteVector(MERGE_MINE_PREFIX);
+        coinbaseBytes.insert(coinbaseBytes.end(), hashChainRoot.begin(),
+                             hashChainRoot.end());
+        std::vector<uint8_t> data = {1, 2, 3, 4, 5, 6, 7, 8};
+        coinbaseBytes.insert(coinbaseBytes.end(), data.begin(), data.end());
+
+        CMutableTransaction tx;
+        tx.vin.resize(1);
+        tx.vin[0].scriptSig =
+            CScript(coinbaseBytes.begin(), coinbaseBytes.end());
+
+        auxpow.coinbaseTx = MakeTransactionRef(tx);
+        auxpow.vMerkleBranch.resize(5);
+        auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+            auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+
+        for (const Consensus::Params &params : allParams) {
+            BOOST_CHECK_EQUAL(
+                ErrorString(auxpow.CheckAuxBlockHash(hashChildBlock, 1, params))
+                    .original,
+                "AuxPow merkle branch size does not match parent coinbase");
+        }
+    }
+
+    // Test every chain index (for merkle height 7, this is 128).
+    // Merge-mining requries us to have nChainIndex to be a specific value based
+    // on the merge-mine nonce, chain ID and the merkle height, so for one of
+    // these values, the check will succeed, and for the others, it will error
+    // with "wrong chain index"
+    for (uint32_t nChainIndex = 0; nChainIndex < 128; ++nChainIndex) {
+        const uint32_t merkleHeight = 7;
+        uint256 hashChildBlock = uint256S(
+            "123456789012345678901234567890123456789012345678901234567890abcd");
+
+        CAuxPow auxpow{};
+        auxpow.nIndex = 0;
+        auxpow.nChainIndex = nChainIndex;
+        auxpow.vChainMerkleBranch.resize(7);
+
+        uint256 hashChainRoot = ComputeMerkleRootForBranch(
+            hashChildBlock, auxpow.vChainMerkleBranch, nChainIndex);
+        std::reverse(hashChainRoot.begin(), hashChainRoot.end());
+
+        std::vector<uint8_t> coinbaseBytes = ToByteVector(MERGE_MINE_PREFIX);
+        coinbaseBytes.insert(coinbaseBytes.end(), hashChainRoot.begin(),
+                             hashChainRoot.end());
+        std::vector<uint8_t> data = {1 << merkleHeight, 0, 0, 0, 5, 6, 7, 8};
+        coinbaseBytes.insert(coinbaseBytes.end(), data.begin(), data.end());
+
+        CMutableTransaction tx;
+        tx.vin.resize(1);
+        tx.vin[0].scriptSig =
+            CScript(coinbaseBytes.begin(), coinbaseBytes.end());
+
+        auxpow.coinbaseTx = MakeTransactionRef(tx);
+        auxpow.vMerkleBranch.resize(5);
+        auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+            auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+
+        for (const Consensus::Params &params : allParams) {
+            const uint32_t expectedChainIndex = CalcExpectedMerkleTreeIndex(
+                0x08070605, AUXPOW_CHAIN_ID, merkleHeight);
+            if (expectedChainIndex == nChainIndex) {
+                BOOST_CHECK(auxpow.CheckAuxBlockHash(hashChildBlock,
+                                                     AUXPOW_CHAIN_ID, params));
+            } else {
+                BOOST_CHECK_EQUAL(
+                    ErrorString(auxpow.CheckAuxBlockHash(
+                                    hashChildBlock, AUXPOW_CHAIN_ID, params))
+                        .original,
+                    "AuxPow wrong chain index");
+            }
+        }
+    }
+}
+
+uint256 GenHash(MMIXLinearCongruentialGenerator &lcg) {
+    uint256 hash;
+    for (uint8_t &byte : hash) {
+        byte = lcg.next();
+    }
+    return hash;
+}
+
+BOOST_AUTO_TEST_CASE(CheckAuxBlockHash_rng_test) {
+    const std::vector<Consensus::Params> allParams{
+        CChainParams::Main({})->GetConsensus(),
+        CChainParams::TestNet({})->GetConsensus(),
+        CChainParams::RegTest({})->GetConsensus()};
+
+    MMIXLinearCongruentialGenerator lcg;
+    CDataStream ss{0, 0};
+    // Randomly generate a lot of configurations and test for successes.
+    for (size_t testCase = 0; testCase < 2048; testCase++) {
+        // Generate random parameters
+        const uint32_t chainId = lcg.next() % 0x10000;
+        const uint32_t parentChainId = (chainId + 1) % 0x10000;
+        const uint32_t chainMerkleHeight = lcg.next() % 31;
+        const uint32_t chainTreeSize = 1 << chainMerkleHeight;
+        const uint32_t nMergeMineNonce = lcg.next();
+        const uint256 hashChildBlock = GenHash(lcg);
+        const uint32_t blockMerkleHeight = lcg.next() % 32;
+        const uint32_t versionLowBits = lcg.next() % 256;
+
+        CAuxPow auxpow{};
+        auxpow.vChainMerkleBranch.resize(chainMerkleHeight);
+        for (uint256 &hash : auxpow.vChainMerkleBranch) {
+            hash = GenHash(lcg);
+        }
+        auxpow.vMerkleBranch.resize(blockMerkleHeight);
+        for (uint256 &hash : auxpow.vMerkleBranch) {
+            hash = GenHash(lcg);
+        }
+
+        auxpow.nIndex = 0;
+        auxpow.nChainIndex = CalcExpectedMerkleTreeIndex(
+            nMergeMineNonce, chainId, chainMerkleHeight);
+
+        uint256 hashChainRoot = ComputeMerkleRootForBranch(
+            hashChildBlock, auxpow.vChainMerkleBranch, auxpow.nChainIndex);
+        std::reverse(hashChainRoot.begin(), hashChainRoot.end());
+
+        std::vector<uint8_t> coinbaseBytes = ToByteVector(MERGE_MINE_PREFIX);
+        coinbaseBytes.insert(coinbaseBytes.end(), hashChainRoot.begin(),
+                             hashChainRoot.end());
+        ss.clear();
+        ss << chainTreeSize;
+        ss << nMergeMineNonce;
+        Span<const uint8_t> span = MakeUCharSpan(ss);
+        coinbaseBytes.insert(coinbaseBytes.end(), span.begin(), span.end());
+
+        CMutableTransaction tx;
+        tx.vin.resize(1);
+        tx.vin[0].scriptSig =
+            CScript(coinbaseBytes.begin(), coinbaseBytes.end());
+
+        auxpow.coinbaseTx = MakeTransactionRef(tx);
+        auxpow.parentBlock.nVersion =
+            MakeVersionWithChainId(parentChainId, versionLowBits);
+        auxpow.parentBlock.hashMerkleRoot = ComputeMerkleRootForBranch(
+            auxpow.coinbaseTx->GetHash(), auxpow.vMerkleBranch, 0);
+
+        for (const Consensus::Params &params : allParams) {
+            util::Result<std::monostate> result =
+                auxpow.CheckAuxBlockHash(hashChildBlock, chainId, params);
+            BOOST_CHECK_MESSAGE(result, ErrorString(result).original);
+        }
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
